src/ripple/basics/SlabAllocator.h

```diff
@@ -57,40 +57,66 @@ class SlabAllocator
         // A linked list of appropriately sized free buffers:
         std::uint8_t* l_ = nullptr;
 
-        // The next memory block
+        // The next block:
         SlabBlock* next_;
 
         // The underlying memory block:
-        std::uint8_t const* const p_ = nullptr;
+        std::atomic<std::uint8_t*> data_ = nullptr;
 
-        // The extent of the underlying memory block:
-        std::size_t const size_;
+        // The number of items allocated from this block
+        std::atomic<std::size_t> avail_{0};
 
-        SlabBlock(
-            SlabBlock* next,
+        SlabBlock(SlabBlock* next) : next_(next)
+        {
+        }
+
+        ~SlabBlock()
+        {
+            // We don't release the allocated memory here: slabs are only
+            // destroyed when the process is shutting down, and we cannot
+            // guarantee that this isn't called before the destruction of
+            // all the objects that are allocated in the buffer we have.
+        }
+
+        std::uint8_t*
+        assign(
```

Comment:
This function both assigns and allocates, maybe rename it?

Reply:
This function doesn't allocate: the memory block is passed _into_ this function.